src/bytecode/scanner.rs

```diff
@@ -1,312 +0,0 @@
-use anyhow::{anyhow, bail, Result};
-
-#[derive(Debug)]
-pub struct Scanner {
-    source: Vec<char>, // TODO: this is not ideal, we should be pointing back to the original source string to save memory
-    start: usize,      // The start of the current lexeme
-    current: usize,    // The current character under consideration
-    line: usize,
-}
-
-impl Scanner {
-    fn new(source: &str) -> Scanner {
-        Scanner {
-            source: source.chars().collect(),
-            start: 0,
-            current: 0,
-            line: 1,
-        }
-    }
-
-    fn make_token(&self, typ: TokenType) -> Token {
-        Token {
-            typ,
-            start: self.start,
-            length: self.current - self.start,
-            line: self.line,
-        }
-    }
-
-    fn is_at_end(&self) -> bool {
-        self.current == self.source.len() - 1
-    }
-
-    fn advance(&mut self) -> char {
-        self.current += 1;
-        self.source[self.current - 1]
-    }
-
-    fn match_char(&mut self, expected: char) -> bool {
-        if self.is_at_end() || self.source[self.current] != expected {
-            false
-        } else {
-            self.current += 1;
-            true
-        }
-    }
-
-    fn peek(&self) -> char {
-        self.source[self.current]
-    }
-
-    fn peek_next(&self) -> Option<char> {
-        if self.is_at_end() {
-            None
-        } else {
-            Some(self.source[self.current + 1])
-        }
-    }
-
-    fn skip_whitespace(&mut self) {
-        loop {
-            if self.is_at_end() {
-                return;
-            }
-
-            match self.peek() {
-                ' ' | '\r' | '\t' => {
-                    self.advance();
-                }
-                '\n' => {
-                    self.line += 1;
-                    self.current += 1;
-                }
-                '/' => {
-                    if let Some('/') = self.peek_next() {
-                        while self.peek() != '\n' && !self.is_at_end() {
-                            self.current += 1;
-                        }
-                    } else {
-                        return;
-                    }
-                }
-                _ => return,
-            }
-        }
-    }
-
-    fn string(&mut self) -> Result<Token> {
-        while self.peek() != '"' && !self.is_at_end() {
-            if self.peek() == '\n' {
-                self.line += 1;
-            }
-            self.advance();
-        }
-
-        if self.is_at_end() {
-            bail!("Unterminated string");
-        }
-
-        self.advance(); // advance over the closing "
-
-        Ok(self.make_token(TokenType::String))
-    }
-
-    fn digit(&mut self) -> Token {
-        while self.peek().is_ascii_digit() {
```

Comment:
Will this loop panic at end of source?

Reply:
Sorry this whole time I didn't realize I was commenting on a deleted file :)

Reply:
:D yeah I think this is one of the reasons I ended up unhappy with this implementation - it didn't really make it clear when we hit end of source. The new implementation should handle that all gracefully (I think!)